Twine 2 JSON

videlais opened the pull request:
I don't love the idea of adding even more formats (Twine 1: TWS, HTML, Twee; Twine 2: Archive, HTML, Twee), but it does feel like this might be needed at some point, if for no other reason than to define it more officially.

videlais commented on twine-2-jsonoutput-doc.md:
```diff
@@ -0,0 +1,61 @@
+  },
+  "zoom": 0.25,
+  "passages": [
+    {
```
Yup. It should. Good catch.

videlais reviewed (commented):
Passages definitely should have `content`.

videlais commented:
> I can't tell from looking at this--is this a proposed new spec, or is this documenting how JSON has been used in the wild?

A new spec. We were talking about multiple tools moving towards JSON representation over a year ago. This would help standardize output between them.

> I think it would be good to get feedback from Emilia Lazer-Walker since they publish [Twison](https://github.com/lazerwalker/twison). It looks like there are some superficial conflicts between this and Twison.

I'm okay with adding in `pid` (for compatibility with Twine 2 HTML) and changing `content` to `text` for passages. I don't think `link` is as useful as it might seem, though. Multiple story formats support link syntax Twine itself doesn't. SugarCube, for example, allows for story variables within links, which would make generating them difficult.

videlais commented:
Hi, @lazerwalker. We are in the process of reviewing a possible JSON Twine specification for tool interoperability. As you created [Twison](https://github.com/lazerwalker/twison), we were wondering if you wanted to weigh in on this specification in case it would impact your work or those using your story format.

As @klembot noted in the above conversation, there are some changes between your last story format update in 2020 and this proposal, notably where this proposal tries to match the Twee 3 specification approved in 2021, but I'm open to trying to more closely match general community usage (changing `content` to `text`, adding `pid`, etc.). I'm against using `link` as a property because link syntax is not standardized in Twine (and current versions of SugarCube and Snowman support syntax Twine itself does not), but we can change other things to bring the newer standard in-line with general usage of JSON with Twine tools.

lazerwalker commented:
Hi! Thanks for reaching out!

For better or for worse, I don't have a lot of insight to add — I know a lot of people continue to use my tool, but I hear very little about what they're using it for. My gut instinct is it's ~fine~ to just stick to the Twee spec and let any users migrating change their own property names, given how superficial the differences are, but that's very much just my personal opinion rather than being informed by having deep knowledge of how people are integrating my tool within their workflows.

(And totally agree about links — it was easy for me to throw that in as an experimental feature of dubious provenance. I wouldn't expect something like that from a formal tool, there's plenty of space for people to build that sort of link parsing themselves on top of official JSON export if they find it valuable)

videlais commented:
> For better or for worse, I don't have a lot of insight to add — I know a lot of people continue to use my tool, but I hear very little about what they're using it for. My gut instinct is it's fine to just stick to the Twee spec and let any users migrating change their own property names, given how superficial the differences are, but that's very much just my personal opinion rather than being informed by having deep knowledge of how people are integrating my tool within their workflows.

Thanks for replying, @lazerwalker! Unfortunately, this is a common problem with many story formats and Twine-compatible tools. We -- well, I can only speak for myself here -- often do not hear anything until there is a major problem or functionality breaks.

> In general, the argument for JSON export over just Twee is using Twine to generate narrative content meant to be embedded within another game. Having to pull in a Twee parser feels extremely heavy compared to just using a standard data format you can already parse. (Should you be encouraging Twine be used that way, rather than reaching for better-suited tools? I could make a reasonable argument either way.)

Ha. You have made the same argument for why many of us worked toward Twee 3 notation over the exclusively of using HTML a few years ago: the parser technical debt cost for tools and game engine integration is higher with HTML because the need to package an additional library and the complications of de-serialization of data from elements and their attributes.

I was initially against including JSON as an encoding format last year because of the amount of labor behind trying to get Twee 3 public, but I agree with you matching a more standardized format would be helpful in most use cases. In fact, based on some recent conversations I've had with folx, I think pushing for partial story encoding might be a good future-facing compromise to let people optionally encode parts of a story using JSON instead of dealing with a common Twine 2 HTML issue of people importing example or tutorial code and then being stuck with its initial IFID.

---

@klembot and @tmedwards: 

I've changed three things in a new commit:
- `content` to `text` for passage data encoding.
- Changed `ifid` to optional for story metadata.
- Added a new section on partial story encoding. One of the experiments I have been doing as part of some academic research this year is having libraries and other tools generate narrative content (think, basically, hundreds of passages). Both Twine 2 HTML and Twee 3 generally represent whole stories. This new paragraph explains how some passages containing content, CSS, or JavaScript code could be encoded and then combined together with other JSON content as "story fragments" to make one single story during compilation. It's my hope this opens the door to plugins and CSS templates stored as JSON.

greyelf commented on twine-2-jsonoutput-doc.md:
```diff
@@ -2,11 +2,11 @@
+To maintain close compatibility with the existing output formats of [Twine 2 HTML](https://github.com/iftechfoundation/twine-specs/blob/master/twine-2-htmloutput-spec.md) and [Twee 3 notation](https://github.com/iftechfoundation/twine-specs/blob/master/twee-3-specification.md), story properties encoded in JSON mirror those found in the [`StoryData` passage in Twee 3 notation](https://github.com/i ...
 
-- `ifid`: (string) Required. Maps to `<tw-storydata ifid>`.
+- `ifid`: (string) Optional. Maps to `<tw-storydata ifid>`.
```
I'm curious, why is the `ifid` **optional** in this JSON related specification when it is **required** in both the [twee 3](https://github.com/iftechfoundation/twine-specs/blob/master/twee-3-specification.md#storydata) and [twine 2 html output](https://github.com/iftechfoundation/twine-specs/blob/master/twine-2-htmloutput-spec.md#story-data) specifications?

videlais replied:
> Both Twine 2 HTML and Twee 3 generally represent whole stories. This new paragraph explains how some passages containing content, CSS, or JavaScript code could be encoded and then combined together with other JSON content as "story fragments" to make one single story during compilation. It's my hope this opens the door to plugins and CSS templates stored as JSON.

To allow for partial story encodings. This is a known problem with both Twine 2 HTML and Twee 3 code and examples. Potentially, this encoding could represent parts or some of a story rather than inherit the IFID same problem.

tmedwards commented:
I hadn't mentioned it prior to now, but the `develop` branch of Tweego already includes a JSON output mode.  Since it's currently unreleased, I was thinking of just conforming to whatever the spec here ends up being.

I'm thinking now however that I should at least compare the two to see if I'd have anything meaningful to make note of.  I tend to doubt it, but it can't hurt to check.

I'll see if I can do that soonish.

tmedwards commented on twine-2-jsonoutput-doc.md:
There should be a root `name` property, for the story name.

videlais replied:
Added. I've also added `creator` and `creator-version` to match Twine 2 HTML attributes.

tmedwards commented:
Per https://github.com/iftechfoundation/twine-specs/pull/24#issuecomment-1756391919: Tweego's pending JSON output held no surprises.

videlais commented:
@greyelf I'd really like for `ifid` to be optional to allow for encoding plugins and CSS templates for others, but this is not a breaking issue for me. If y'all feel strong enough it should be required, I'll update.

---

@tmedwards In the latest commit, `name`, `creator`, and `creator-version` were added along with the same text from the Twine 2 HTML spec.

I have JSON support in the development version of Extwee as well. Once this gets approved, I'll hopefully be releasing an update for Extwee soon after to match this along with the other proposed documentation.

---

@klembot It is my hope, assuming `ifid` remains optional, this spec could open the door to plugins, CSS templates, and other things for people to "import" existing passages into a story. Obviously, for Twine support, this would take months to implement, but this could be a step toward that. If my current research funding lasts into mid-2024, I'd be happy to help. If not, well, maybe in the 15-20 year span of Twine's life it might get plugin support.

tmedwards commented:
Something just occurred to me.  How are the script and stylesheet sections going to be handled?

I assume as their own properties?  E.g., root `script` and `style`.

videlais commented:
> I assume as their own properties? E.g., root `script` and `style`.

My original thinking was to use `script` and `style` tags on individual passages like the following:

```json
{
  "name": "Minimum CSS Template Example",
  "passages": 
  [
   {
     "name": "CSS Code",
     "tags": ["style"],
     "text": "#body{color:red;}"
   } 
  ]
}
```

Would it make more sense to have them as root properties?

- `style`: (string) Optional. Story Stylesheet content. Maps to `<style role="stylesheet" id="twine-user-stylesheet" type="text/twine-css"></style>`.
- `script`: (string) Optional. Story JavaScript content. Maps to `<script role="script" id="twine-user-script" type="text/twine-javascript"></script>`.

hituro commented:
Not sure if this is a relevant concern, but mapping a `[script]` or `[style]` passage to a top-level property would make it impossible to convert back and forth between output formats in a stable way. e.g. a Twee file containing a `[script]` tagged passage only outputs to the story JS block when compiled. Converting from twee to JSON and back would then not restore the original passages.

videlais commented:
> Converting from twee to JSON and back would then not restore the original passages.

Correct. As tmedwards points out, moving from Twee into JSON would lose passages. The same is also true of moving from Twee to Twine 2 HTML. Matching the Twine 2 HTML spec means collapsing down to its single `<style>` and `<script>` elements.

However, assuming JSON gets approved, you could combine Twee and JSON content, @hituro. Potentially, at least as of this writing, a story could be written in Twee using a number of external tools and the story compilation tool given JSON as additional sources of other passages, CSS, or JavaScript.

Something I keep pushing for with this specification is to quickly serialize passages and stories without them needing to be "complete" stories with an IFID. This would, at least I hope, open the door for many things including CSS templates for story formats, easily importing and sharing tutorial examples (without all of them having the same IFID), and allowing people to quickly load plugins and external JavaScript libraries with Twine projects.

---

> I'd expect this to mirror the Twine 2 data chunk as much as is sensible.

Agreed, @tmedwards. `style` and `script` root properties added via latest commit.

videlais commented:
The last outstanding issue is the requirement of an `ifid`. Do you feel strongly I should change the field from "Optional" to "Required", @greyelf? 

If this is the only obstacle for acceptance, I will revise the spec to match. I'd like for JSON to be able to represent partial and in-progress stories, as it would make serializing data out and potentially accepting new data into ongoing stories or as part of the compilation process easier for JavaScript-based processes, but if this is the only thing left, I will make the change and we can revisit the concept of plugins/templates in the future.

greyelf commented:
> Do you feel strongly I should change the field from "Optional" to "Required"

@videlais. I don't have strong feelings either way, however...

> and potentially accepting new data into ongoing stories

@klembot 
Correct me if I'm wrong, but isn't the IFID of a Project/Story html file the main unique identifier of said file, and the IFID of an imported Story used to determine if the content being imported is meant to:
* replace that of an existing Project
* be added as a new Project

And if that is the case, then how will an "import from JSON" option know which Project to add the contents of the JSON file to when the file doesn't include a IFID?, especially when the Title/Name of a Project is basically meaningless for such programmatically identification purposes?

Or will such an "import from JSON" feature only be the domain of command line tools, where the target project is specified by its file-name?

videlais commented:
@greyelf This is, as far as I know, the first time we have discussed a format not currently used. In all other cases, Twine formats existed and we were documenting how it worked or standardizing community usage (i.e. Twee 3). In this case, we are future-facing something Twine-compatible tools are implementing first with potentially Twine adopting it some time in the future.

> And if that is the case, then how will an "import from JSON" option know which Project to add the contents of the JSON file to when the file doesn't include a IFID?, especially when the Title/Name of a Project is basically meaningless for such programmatically identification purposes?

Potentially, namespaces could be used. This definitely needs to be addressed before it becomes a central part of plugin/template usage in Twine itself. I totally agree. But, like klembot mentions, we aren't there yet.

> Or will such an "import from JSON" feature only be the domain of command line tools, where the target project is specified by its file-name?

Right now, just Twine-compatible tools. Maybe Twine adopts it in future, maybe not, but this is an exciting development in the space where tool-interoperability is a focus _before_ Twine uses a format instead of trying to standardize some amount of time after it has been used.

---

> I'm not making a commitment right now to add JSON import to Twine. I'm open to exploring the plugin/extension functionality Dan is discussing in this PR, but there are a lot of implementation details to sort out.

Right. That's a good point, @klembot. I have been getting ahead of myself some in my discussion. Right now, it's only a format. Assuming it gets approved, Tweego and Extwee will support it as an input and output format when new versions come out.

Potentially, if there is strong interest, and at least from my perspective, I have funding through next summer, I'd like to work toward seeing if this format *could* support a plugin-like system for Twine and compatible tools. That's future work, and might require a totally different format, to be honest. My feeling is JSON could work, but without more research and labor toward that end, I'm not sure.

videlais commented:
@greyelf Is there any other feedback you want to incorporate or changes needed? I agree with both you and klembot that more open discussion and planning is needed before this could be used for plugins for Twine itself, and am hopeful we can all continue that conversation into early 2024 (at which point I may need to go quiet for awhile again).

greyelf commented:
@videlais My concerns have been resolved.
